lib/libutee/image.c

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2016, Linaro Limited
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <stdlib.h>
+#include <utee_defines.h>
+#include "image.h"
+
+static uint32_t color_to_pixel(uint32_t color)
+{
+	/* Convert from ARGB to RGBA and byte swap */
+	return TEE_U32_BSWAP((color >> 24) | (color << 8));
+}
+
+struct image *image_alloc(size_t width, size_t height, uint32_t color)
+{
+	struct image *image = malloc(sizeof(*image));
+	size_t n;
+	uint32_t *b;
+
+	if (!image)
+		return NULL;
+	image->blen = sizeof(uint32_t) * height * width;
+	image->buf = malloc(image->blen);
+	if (!image->buf) {
+		free(image);
+		return NULL;
+	}
+	b = image->buf;
+	image->height = height;
+	image->width = width;
+	for (n = 0; n < (height * width); n++)
+		b[n] = color_to_pixel(color);
+	return image;
+}
+
+void image_free(struct image *image)
+{
+	if (image) {
+		free(image->buf);
+		free(image);
+	}
+}
+
+void *image_get_pixel_ptr(struct image *image, size_t x, size_t y)
+{
+	uint32_t *b = image->buf;
+	size_t pos;
+
+	if (x >= image->width || y >= image->height)
```

Comment:
It might be that I don't fully follow the buffer sizes etc, but should it be `>=` on the `y` check here? Wouldn't `>` be sufficient? With `>=` don't you loose the last line?